scripts/dev/upstream_merge/upstream_merge.sh

```diff
@@ -0,0 +1,206 @@
+#!/bin/bash
+
+# This script can be used for merging latest upstream for each of the
+# repos described in repos.conf
+#
+# For each local repo, it does following
+# 1. Check if local_base_branch specified in repos.conf exists
+# 2. Checkout local_base_branch and run git pull
+# 3. Add a 'remote' named 'automerge_upstream'
+#    a. If the remote already exists, remove it first
+# 4. Pulls remote branch into a branch named 'dev/automerge/upstream/<upstream_branch>'
+#    a. If the branch name already exists, remove it first
+# 5. Creates a branch named 'dev/automerge/ni' from 'nilrt/master/hardknott' and checkout
+#    a. If the branch name exists, ask to delete or skip or cancel
+# 6. Merges 'dev/automerge/upstream' into 'dev/automerge/ni'
+#    a. Success or failure is reported; script continues in either case
+
+DIR=$(dirname ${BASH_SOURCE})
+CONF_FILE="$DIR/repos.conf"
+
+REMOTE_REPO_NAME="automerge_upstream"
+LOCAL_UPSTREAM_BRANCH_PREFIX="dev/automerge/upstream"
+LOCAL_BRANCH_NAME="dev/automerge/ni"
+
+usage() {
+   echo "Usage: $0 [-c <conf file>][-h] " 1>&2
+   exit 1
+}
+
+parse_args() {
+   local OPTIND o
+   while getopts "c:h" o; do
+      case "${o}" in
+         c)
+            CONF_FILE=${OPTARG}
+            ;;
+         h)
+            usage
+            ;;
+         *)
+            usage
+            ;;
+      esac
+   done
+   shift $((OPTIND-1))
+}
+
+sanity_test_repo() {
+   LOCAL_BASE_BRANCH=$1
+   if ! `git rev-parse --verify $LOCAL_BASE_BRANCH &> /dev/null`; then
+      echo ""
+      echo "    Branch $LOCAL_BASE_BRANCH does not exist. Exiting"
+      exit 1
+   fi
+}
+
+update_local_base_branch() {
+   LOCAL_BASE_BRANCH=$1
+   if `git checkout $LOCAL_BASE_BRANCH &> /dev/null`; then
+      if ! `git pull &> /dev/null`; then
+         echo ""
+         echo "    Error pulling latest on $LOCAL_BASE_BRANCH. Exiting"
+         exit 1
+      fi
+   else
+      # Error changing branch
+      echo ""
+      echo "    Error switching to branch $LOCAL_BASE_BRANCH. Exiting"
+      exit 1
+   fi
+}
+
+add_remote() {
+   UPSTREAM_REPO=$1
+
+   git remote remove $REMOTE_REPO_NAME &> /dev/null || true
+   git remote add $REMOTE_REPO_NAME $UPSTREAM_REPO
+}
+
+sync_remote_branch() {
+   UPSTREAM_BRANCH=$1
+   LOCAL_UPSTREAM_BRANCH=$2
+
+   git branch -D $LOCAL_UPSTREAM_BRANCH &> /dev/null || true
+   if ! `git fetch $REMOTE_REPO_NAME $UPSTREAM_BRANCH:$LOCAL_UPSTREAM_BRANCH &> /dev/null`; then
```

Comment:
Is any of this branch manipulation actually necessary to create a rebased branch? I don't think it is.

Reply:
Are you suggesting we don't create `LOCAL_UPSTREAM_BRANCH` but just fetch `REMOTE_REPO_NAME/UPSTREAM_BRANCH` and use it directly?

Reply:
Got rid of `LOCAL_UPSTREAM_BRANCH` as discussed